File: LifeCounterApi/Models/DTOs/UtilizadorDto.cs
using System.ComponentModel.DataAnnotations;

namespace LifeCounterApi.Models.DTOs;

public class UtilizadorDto
{
    public Guid Id { get; set; }
    public string Nome { get; set; } = string.Empty;
    public string Email { get; set; } = string.Empty;
    public DateTime CriadoEm { get; set; }
    public DateTime AtualizadoEm { get; set; }
}

public class UtilizadorCreateDto
{
    [Required(ErrorMessage = "O nome é obrigatório")]
    [MaxLength(200, ErrorMessage = "O nome não pode ter mais de 200 caracteres")]
    public string Nome { get; set; } = string.Empty;
    
    [Required(ErrorMessage = "O email é obrigatório")]
    [EmailAddress(ErrorMessage = "Email inválido")]
    [MaxLength(200, ErrorMessage = "O email não pode ter mais de 200 caracteres")]
    public string Email { get; set; } = string.Empty;
}

public class UtilizadorUpdateDto
{
    [MaxLength(200, ErrorMessage = "O nome não pode ter mais de 200 caracteres")]
    public string? Nome { get; set; }
    
    [EmailAddress(ErrorMessage = "Email inválido")]
    [MaxLength(200, ErrorMessage = "O email não pode ter mais de 200 caracteres")]
    public string? Email { get; set; }
}

File: life-counter-api-rust/src/dto/utilizador_dto.rs
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use validator::Validate;

#[derive(Debug, Serialize, Deserialize)]
pub struct UtilizadorDto {
    pub id: Uuid,
    pub nome: String,
    pub email: String,
    pub criado_em: NaiveDateTime,
    pub atualizado_em: NaiveDateTime,
}

#[derive(Debug, Deserialize, Validate)]
pub struct UtilizadorCreateDto {
    #[validate(length(min = 1, max = 200, message = "O nome é obrigatório e não pode ter mais de 200 caracteres"))]
    pub nome: String,
    
    #[validate(email(message = "Email inválido"))]
    #[validate(length(min = 1, max = 200, message = "O email é obrigatório e não pode ter mais de 200 caracteres"))]
    pub email: String,
}

#[derive(Debug, Deserialize, Validate)]
pub struct UtilizadorUpdateDto {
    #[validate(length(max = 200, message = "O nome não pode ter mais de 200 caracteres"))]
    pub nome: Option<String>,
    
    #[validate(email(message = "Email inválido"))]
    #[validate(length(max = 200, message = "O email não pode ter mais de 200 caracteres"))]
    pub email: Option<String>,
}
